### PR TITLE
Create OptionalEntity to allow configurable addition of child entities

### PR DESCRIPTION
**Pull request by grkvlt**
Allows entity creation to be controlled using a configuration key value.

### CODE REVIEW

File: core/src/main/java/org/apache/brooklyn/entity/stock/OptionalEntity.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+ *       $brooklyn:entitySpec:
+ *         type: load-balancer
+ *         brooklyn.config:
+ *           proxy.port: 8080
+ *           loadbalancer.serverpool: $brooklyn:entity("servers")
+ * </pre>
+ */
+@Beta
+@ImplementedBy(OptionalEntityImpl.class)
+public interface OptionalEntity extends BasicStartable {
```
Comment (neykov) [inline core/src/main/java/org/apache/brooklyn/entity/stock/OptionalEntity.java]: Would a `ConditionalEntity` be a better name?

File: core/src/main/java/org/apache/brooklyn/entity/stock/OptionalEntityImpl.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+    @Override
+    public void start(Collection<? extends Location> locations) {
+        Boolean create = config().get(CREATE_OPTIONAL_ENTITY);
+        if (Boolean.TRUE.equals(create)) {
+            EntitySpec<?> spec = Preconditions.checkNotNull(config().get(OPTIONAL_ENTITY_SPEC));
+            addChild(EntitySpec.create(spec));
+        }
+        super.start(locations);
+    }
```
Comment (neykov) [inline core/src/main/java/org/apache/brooklyn/entity/stock/OptionalEntityImpl.java]: Check if the entity is already created - to handle start -> stop -> start/restart.

Comment (grkvlt) [inline core/src/main/java/org/apache/brooklyn/entity/stock/OptionalEntityImpl.java]: Good point.

File: core/src/test/java/org/apache/brooklyn/entity/stock/OptionalEntityTest.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+import org.apache.brooklyn.api.entity.EntitySpec;
+import org.apache.brooklyn.api.location.LocationSpec;
+import org.apache.brooklyn.api.mgmt.ManagementContext;
+import org.apache.brooklyn.core.entity.Entities;
+import org.apache.brooklyn.core.location.SimulatedLocation;
+import org.apache.brooklyn.core.test.entity.LocalManagementContextForTests;
+import org.apache.brooklyn.core.test.entity.TestApplication;
+import org.apache.brooklyn.core.test.entity.TestEntity;
+
+public class OptionalEntityTest {
```
Comment (aledsage-tmp) [inline core/src/test/java/org/apache/brooklyn/entity/stock/OptionalEntityTest.java]: You can extend BrooklynAppUnitTestSupport to save you writing a few lines here.

**Comment by aledsage-tmp** [comment]
@grkvlt agree with @neykov about the name ConditionalEntity. And similar rename in config keys. Also worth adding test for restart.

**Comment by grkvlt** [comment]
@aledsage-tmp @neykov I changed the name and made the suggested changes

**Comment by neykov** [comment]
Thanks @grkvlt, merging.

**Pull request merged**